gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/MultiHopsFlowToJobSpecCompiler.java

```diff
@@ -0,0 +1,389 @@
+
+import avro.shaded.com.google.common.annotations.VisibleForTesting;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
+import com.typesafe.config.Config;
+
+import org.apache.gobblin.service.modules.utils.DistancedNode;
+import org.apache.commons.lang3.StringUtils;
+import org.jgrapht.graph.DirectedWeightedMultigraph;
+import org.slf4j.Logger;
+import org.apache.gobblin.runtime.api.FlowEdge;
+import org.apache.gobblin.runtime.api.ServiceNode;
+import org.apache.gobblin.runtime.api.FlowSpec;
+import org.apache.gobblin.instrumented.Instrumented;
+import org.apache.gobblin.runtime.api.Spec;
+import org.apache.gobblin.runtime.api.TopologySpec;
+import org.apache.gobblin.service.ServiceConfigKeys;
+import org.apache.gobblin.runtime.spec_executorInstance.BaseServiceNodeImpl;
+import org.apache.gobblin.runtime.api.JobSpec;
+import org.apache.gobblin.runtime.api.JobTemplate;
+import org.apache.gobblin.runtime.api.SpecExecutor;
+import org.apache.gobblin.runtime.api.SpecNotFoundException;
+import org.apache.gobblin.runtime.job_spec.ResolvedJobSpec;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+// Users are capable to inject hints/prioritization into route selection, in two forms:
+// 1. PolicyBasedBlockedConnection: Define some undesired routes
+// 2. Specified a complete path. FlowCompiler is responsible to verify if the path given is valid.
+
+@Slf4j
+public class MultiHopsFlowToJobSpecCompiler extends BaseFlowToJobSpecCompiler {
+
+  /* TODO:
+    1. When the connection between hops is no longer configuration based,
+     should ensure the updates of following two data structures atomic with TopologySpec Catagory updates
+    2. Figure out a way to inject weight information in Topology. We provide listenable interface in weightGraph, so
+     those changes in topology should reflect in weightGraph as well.
+  */
+
+  @Getter
+  private DirectedWeightedMultigraph<ServiceNode, FlowEdge> weightedGraph =
+      new DirectedWeightedMultigraph<>(LoadBasedFlowEdgeImpl.class);
+
+  //Contains the user-specified connection that are not desired to appear in data movement path.
+  //It can be used for avoiding known expensive or undesired data movement.
+  public Optional<Multimap<String, String>> optionalPolicyBasedBlockedConnection;
+
+  // Contains user-specified complete path of how the data movement is executed from source to sink.
+  private Optional<String> optionalUserSpecifiedPath;
```

Comment:
Currently it is a comma-separated string and is transformed into a list of path in `userSpecifiedPathVerificator`. I think it may be easier to configure for user.